use_cases/eluc/predictors/neural_network/neural_net_predictor.py

```diff
@@ -0,0 +1,297 @@
+    """
+    def __init__(self, X: torch.FloatTensor, y: torch.LongTensor):
+        super().__init__()   
+        self.X = torch.tensor(X, dtype=torch.float32)
+        self.y = torch.tensor(y)
+
+    def __len__(self):
+        return len(self.X)
+
+    def __getitem__(self, idx: int) -> tuple:
+        return self.X[idx], self.y[idx]
+
+
+class ELUCNeuralNet(torch.nn.Module):
+    """
+    Custom torch neural network module.
+    :param in_size: number of input features
+    :param hidden_sizes: list of hidden layer sizes
+    :param linear_skip: whether to concatenate input to hidden layer output
+    :param dropout: dropout probability
+    """
+    class EncBlock(torch.nn.Module):
+        """
+        Encoding block for neural network.
+        Simple feed forward layer with ReLU activation and optional dropout.
+        """
+        def __init__(self, in_size: int, out_size: int, dropout: float):
+            super().__init__()
+            self.model = torch.nn.Sequential(
+                torch.nn.Linear(in_size, out_size),
+                torch.nn.ReLU(),
+                torch.nn.Dropout(p=dropout)
+            )
+        def forward(self, X: torch.FloatTensor) -> torch.FloatTensor:
+            """
+            Passes input through the block.
+            """
+            return self.model(X)
+
+    def __init__(self, in_size: int, hidden_sizes: list[str], linear_skip: bool, dropout: float):
+        super().__init__()
+        self.linear_skip = linear_skip
+        hidden_sizes = [in_size] + hidden_sizes
+        enc_blocks = [self.EncBlock(hidden_sizes[i], hidden_sizes[i+1], dropout) for i in range(len(hidden_sizes) - 1)]
+        self.enc = torch.nn.Sequential(*enc_blocks)
+        out_size = hidden_sizes[-1] + in_size if linear_skip else hidden_sizes[-1]
+        self.linear = torch.nn.Linear(out_size, 1)
+
+    def forward(self, X: torch.FloatTensor) -> torch.FloatTensor:
+        """
+        Performs a forward pass of the neural net.
+        :param X: input data
+        :return: output of the neural net
+        """
+        hid = self.enc(X)
+        if self.linear_skip:
+            hid = torch.concatenate([hid, X], dim=1)
+        out = self.linear(hid)
+        return out
+
+
+class NeuralNetPredictor(Predictor):
+    """
+    Simple feed-forward neural network predictor implemented in PyTorch.
+    Has the option to use wide and deep, concatenating the input to the output of the hidden layers
+    in order to take advantage of the linear relationship in the data.
+    Data is automatically standardized and the scaler is saved with the model.
+    """
+    def __init__(self, features=None, label=None, hidden_sizes=[4096], linear_skip=True, dropout=0, device="mps",
+            epochs=3, batch_size=2048, optim_params={}, train_pct=1, step_lr_params={"step_size": 1, "gamma": 0.1}): 
+        # Model setup params
+        self.model = None
+        self.features = features
+        self.label = label
```

Comment:
We make sure to keep track of the features and label for when we load the model later